diff --git a/src/rime/context.h b/src/rime/context.h
index 6f514320..17e58051 100644
--- a/src/rime/context.h
+++ b/src/rime/context.h
@@ -79,11 +79,13 @@ class RIME_DLL Context {
   // options and properties starting with '_' are local to schema;
   // others are session scoped.
   void ClearTransientOptions();
+  void AbortCompositionNotification() { abort_notifier_(this); }
 
   Notifier& commit_notifier() { return commit_notifier_; }
   Notifier& select_notifier() { return select_notifier_; }
   Notifier& update_notifier() { return update_notifier_; }
   Notifier& delete_notifier() { return delete_notifier_; }
+  Notifier& abort_notifier() { return abort_notifier_; }
   OptionUpdateNotifier& option_update_notifier() {
     return option_update_notifier_;
   }
@@ -106,6 +108,7 @@ class RIME_DLL Context {
   Notifier select_notifier_;
   Notifier update_notifier_;
   Notifier delete_notifier_;
+  Notifier abort_notifier_;
   OptionUpdateNotifier option_update_notifier_;
   PropertyUpdateNotifier property_update_notifier_;
   KeyEventNotifier unhandled_key_notifier_;
diff --git a/src/rime/service.cc b/src/rime/service.cc
index e40493ae..ad62a5cf 100644
--- a/src/rime/service.cc
+++ b/src/rime/service.cc
@@ -46,6 +46,7 @@ void Session::ClearComposition() {
   if (!engine_)
     return;
   engine_->context()->Clear();
+  engine_->context()->AbortCompositionNotification();
 }
 
 void Session::ApplySchema(Schema* schema) {
